src/main/java/seedu/address/model/ModelManager.java

```diff
@@ -28,7 +31,15 @@
     private final FilteredList<FloatingTask> filteredFloatingTasks;
     private final FilteredList<EventTask> filteredEventTasks;
     private final FilteredList<DeadlineTask> filteredDeadlineTasks;
-
+   
+    //for undo
+    public Stack<TaskBook> stateStack=new Stack<TaskBook>();//a stack of past TaskBook states
+    public final Stack<Command> modifyingDataCommandHistory=new Stack<Command>();
```

Comment:
Ugh, why two stacks that need to be kept in sync? What happens if we need to add more information with each state? Add another stack? A single stack with a class that stores the info you need would be much more appropriate here.